File: Java/Leetcode/LinkedList/RotateList.java
package com.brianprojects.Leetcode.LinkedList;

public class RotateList {
    //Time Complexity: O(n) || Space Complexity: O(n)
    public ListNode rotateRight(ListNode head, int k){
        if(head ==null || k == 0){return head;}

        ListNode current = head;
        int size = 1;
        while(current.next != null){
            current = current.next;
            size++;
        }
        current.next = head;
        current = head;
        ListNode[] array = new ListNode[size];
        for(var i = 0; i<array.length; i++){
            array[(i+k)%size] = current;
            current = current.next;
        }
        ListNode tail = array[array.length-1];
        tail.next = null;

        return array[0];
    }
}

File: Python/Leetcode_Python/LinkedList/RotateList.py
from typing import Optional
from SinglyLL import ListNode


def rotateRight(head:Optional[ListNode], k:int)-> Optional[ListNode]:
    if not head:
        return None
    if not head.next:
        return head

    size = 0

    curr = head

    while curr:
        size+=1
        curr = curr.next

    node = k%size
    if node == 0:
        return head
    
    stopper = size

    p1 = None
    p2 = head
    while stopper != node:
        p1 = p2
        p2 = p2.next
        stopper-=1

    p1.next = None
    curr = p2

    while curr.next:
        curr = curr.next
    curr.next = head
    head = p2
    return head
        
